changelog_unreleased/javascript/pr-7111.md

```diff
@@ -0,0 +1,43 @@
+<!--
+
+1. Choose a folder based on which language your PR is for.
+
+   - For JavaScript, choose `javascript/` etc.
+   - For TypeScript specific syntax, choose `typescript/`.
+   - If your PR applies to multiple languages, such as TypeScript/Flow, choose one folder and mention which languages it applies to.
+
+2. In your chosen folder, create a file with your PR number: `pr-XXXX.md`. For example: `typescript/pr-6728.md`.
+
+3. Copy the content below and paste it in your new file.
+
+4. Fill in a title, the PR number and your user name.
+
+5. Optionally write a description. Many times it’s enough with just sample code.
+
+6. Change ```jsx to your language. For example, ```yaml.
+
+7. Change the `// Input` and `// Prettier` comments to the comment syntax of your language. For example, `# Input`.
+
+8. Choose some nice input example code. Paste it along with the output before and after your PR.
+
+-->
+
+#### Break lines before binary operators ([#7111](https://github.com/prettier/prettier/pull/7111) by [@btmills](https://github.com/btmills))
+
+<!-- prettier-ignore -->
+```jsx
+// Input
+var a = Math.random() * (yRange * (1 - minVerticalFraction)) + minVerticalFraction * yRange - offset;
+
+// Prettier stable
+var a =
+  Math.random() * (yRange * (1 - minVerticalFraction)) +
+  minVerticalFraction * yRange -
+  offset;
+
+// Prettier master
+var a
+  = Math.random() * (yRange * (1 - minVerticalFraction))
```

Comment:
Are we really sure we want to do this with assignments?

Reply:
I started with the most straightforward change in hopes of reducing friction, but a few of the comments on #3806 also prefer assignments at the end, so I’m happy to change it if you’d like.

Reply:
If prettier is in favor of operators first formatting (I’m very supportive of this change), one cool thing about putting assignment operators first is that it makes assignments with several binary operators line up in a really satisfying way:

```js
const isEventTarget
  = value != null
  && typeof value.addEventListener === "function"
  && typeof value.removeEventListener === "function"
  && typeof value.dispatchEvent === "function";
```

It might look foreign at first but you get used to it immediately, and it looks a lot nicer than having the first operand on its own line but indented, IMO. It looks even better in the case of chained assignment. Just try it out next time you find yourself writing a long variable declaration with binary operators and you’ll see what I mean.

Reply:
How can a decision be made here? There is 47/4 voting on @j-f1's [comment](https://github.com/prettier/prettier/issues/3806#issuecomment-360099484), for what that's worth.

You could change it, and see what feedback is received from pre-release versions.

Reply:
This should probably apply to all [assignment operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators) (break after, not before).

Reply:
It turns out the example here was incorrect, and I corrected it in ba9d939. What I didn't know until I started digging in this evening is that assignment expressions are handled separately, so with this change as-is, assignment operators remain on the previous line, and this change only move binary operators to the beginning of lines!

To verify this, I added explicit tests for the example code here with all of the assignment operators in 57e1bda that show all operators at the end of lines before the logic change and [updated the assignment tests in `32e565d`](https://github.com/prettier/prettier/pull/7111/commits/32e565df9ffe42e50bbdf24411258be9fb342c39#diff-f6af98e5ecd0e5be31b254192fa17b49L90-R141) to show that this change only moves the binary operators and leaves the assignments where they were.